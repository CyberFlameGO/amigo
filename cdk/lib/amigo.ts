import path from "path";
import type { CfnRole } from "@aws-cdk/aws-iam";
import { Effect, Policy, PolicyStatement, Role } from "@aws-cdk/aws-iam";
import type { Bucket } from "@aws-cdk/aws-s3";
import { CfnInclude } from "@aws-cdk/cloudformation-include";
import type { App } from "@aws-cdk/core";
import { Stage } from "@guardian/cdk/lib/constants";
import type { GuStackProps, GuStageParameter } from "@guardian/cdk/lib/constructs/core";
import { GuDistributionBucketParameter, GuStack } from "@guardian/cdk/lib/constructs/core";
import { AppIdentity } from "@guardian/cdk/lib/constructs/core/identity";
import {
  GuAllowPolicy,
  GuAnghammaradSenderPolicy,
  GuDescribeEC2Policy,
  GuGetDistributablePolicy,
  GuLogShippingPolicy,
  GuSSMRunCommandPolicy,
} from "@guardian/cdk/lib/constructs/iam";
import { GuS3Bucket } from "@guardian/cdk/lib/constructs/s3";

const yamlTemplateFilePath = path.join(__dirname, "../../cloudformation.yaml");

export class AmigoStack extends GuStack {
  private static app: AppIdentity = {
    app: "amigo",
  };

  private readonly dataBucket: Bucket;

  private get appPolicy(): Policy {
    return new Policy(this, "AppPolicy", {
      policyName: "app-policy",
      statements: [
        /*
        Permissions to enable listing of installed packages created during a bake
        See https://github.com/guardian/amigo/pull/395
         */
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:GetObject"],
          resources: [`${this.dataBucket.bucketArn}/*`],
        }),
      ],
    });
  }

  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const importBucketName = this.withStageDependentValue({
      variableName: "DataBucketName",
      stageValues: {
        [Stage.CODE]: "amigo-data-code",
        [Stage.PROD]: "amigo-data-prod",
      },
    });

    this.dataBucket = new GuS3Bucket(this, "AmigoDataBucket", {
      bucketName: importBucketName,
      existingLogicalId: {
        logicalId: "AmigoDataBucket",
        reason: "To prevent orphaning of the YAML defined bucket",
      },
    });

    const yamlDefinedStack = new CfnInclude(this, "YamlTemplate", {
      templateFile: yamlTemplateFilePath,

      // These override like-named parameters in the YAML template.
      // TODO remove the parameter from the YAML template once each resource that uses it has been CDK-ified.
      parameters: {
        Stage: this.getParam<GuStageParameter>("Stage"), // TODO `GuStageParameter` could be a singleton to simplify this
        DistributionBucketName: GuDistributionBucketParameter.getInstance(this).valueAsString,
      },
    });

    // See https://docs.aws.amazon.com/cdk/latest/guide/use_cfn_template.html#use_cfn_template_cfninclude_access
    const cfnRootRole = yamlDefinedStack.getResource("RootRole") as CfnRole;
    const rootRole = Role.fromRoleArn(this, "RootRole", cfnRootRole.attrArn);

    const ssmPolicy = GuSSMRunCommandPolicy.getInstance(this);

    // TODO Can GuSSMRunCommandPolicy expose its default `PolicyStatement` to easily add actions to it?
    ssmPolicy.addStatements(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: ["*"],

        // required to allow Packer to run SSM commands
        // see https://github.com/guardian/amigo/pull/526 and https://github.com/guardian/amigo/pull/538
        actions: ["ssm:StartSession", "ssm:TerminateSession"],
      })
    );

    const policiesToAttachToRootRole: Policy[] = [
      ssmPolicy,
      GuLogShippingPolicy.getInstance(this),
      new GuGetDistributablePolicy(this, AmigoStack.app),
      new GuAllowPolicy(this, "PackerPolicy", {
        policyName: "packer-required-permissions",
        resources: ["*"],
        actions: [
          "ec2:AttachVolume",
          "ec2:AuthorizeSecurityGroupIngress",
          "ec2:CopyImage",
          "ec2:CreateImage",
          "ec2:CreateKeypair",
          "ec2:CreateSecurityGroup",
          "ec2:CreateSnapshot",
          "ec2:CreateTags",
          "ec2:CreateVolume",
          "ec2:DeleteKeypair",
          "ec2:DeleteSecurityGroup",
          "ec2:DeleteSnapshot",
          "ec2:DeleteVolume",
          "ec2:DeregisterImage",
          "ec2:DescribeImageAttribute",
          "ec2:DescribeImages",
          "ec2:DescribeInstances",
          "ec2:DescribeRegions",
          "ec2:DescribeSecurityGroups",
          "ec2:DescribeSnapshots",
          "ec2:DescribeSubnets",
          "ec2:DescribeTags",
          "ec2:DescribeVolumes",
          "ec2:DetachVolume",
          "ec2:GetPasswordData",
          "ec2:ModifyImageAttribute",
          "ec2:ModifyInstanceAttribute",
          "ec2:ModifySnapshotAttribute",
          "ec2:RegisterImage",
          "ec2:RunInstances",
          "ec2:StopInstances",
          "ec2:TerminateInstances",
          "iam:PassRole",
        ],
      }),
      GuDescribeEC2Policy.getInstance(this),
      GuAnghammaradSenderPolicy.getInstance(this),
      this.appPolicy,
    ];

    policiesToAttachToRootRole.forEach((policy) => policy.attachToRole(rootRole));

    /*
    Looks like some @guardian/cdk constructs are not applying the App tag.
    I suspect since https://github.com/guardian/cdk/pull/326.
    Until that is fixed, we can safely, manually apply it to all constructs in tree from `this` as it's a single app stack.
    TODO: remove this once @guardian/cdk has been fixed.
    */
    AppIdentity.taggedConstruct(AmigoStack.app, this);
  }
}
