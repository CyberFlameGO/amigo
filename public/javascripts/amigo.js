'use strict';

function makeRoleVariablesVisibilityToggleable() {
  $('input.show-role-variables').change(function() {
    var $this = $(this);
    var roleName = $this.data('role');
    var roleVariablesInput = $('input#role-'+roleName+'-variables');
    var checked = $this.prop('checked');
    if (!!checked)
      roleVariablesInput.css('visibility', 'visible');
    else
      roleVariablesInput.css('visibility', 'hidden');
  });
}

function enablePostLinks() {
  $('a.post').click(function() {
    var href = $(this).data('href');
    $('<form action="'+href+'" method="post"></form>').appendTo('body').submit();
  });
}

function relocateStickyComponents() {
  var windowTop = $(window).scrollTop();
  $('.sticky').each(function () {
    var $this = $(this);
    var $anchor = $this.prev(); // the component that you want to make sticky must be preceded by an empty div
    var componentTop = $anchor.offset().top;
    if (windowTop > componentTop - 60) {
      $this.addClass('stick');
      $anchor.height($this.outerHeight());
    } else {
      $this.removeClass('stick');
      $anchor.height(0);
    }
  });
}

function initStickyComponents() {
  $(window).scroll(relocateStickyComponents);
  relocateStickyComponents();
}

$(function(){
  enablePostLinks();
  makeRoleVariablesVisibilityToggleable();
  initStickyComponents();
});
