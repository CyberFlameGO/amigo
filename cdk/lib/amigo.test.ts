import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { AmigoStack } from "./amigo";

describe("The Amigo stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new AmigoStack(app, "AMIgo", {
      stack: "deploy",
      migratedFromCloudFormation: true,
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
