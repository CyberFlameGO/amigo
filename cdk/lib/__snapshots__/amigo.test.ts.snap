// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Amigo stack matches the snapshot 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "AMIgo, an AMI bakery",
  "Mappings": Object {
    "stagemapping": Object {
      "CODE": Object {
        "DataBucketName": "amigo-data-code",
      },
      "PROD": Object {
        "DataBucketName": "amigo-data-prod",
      },
    },
  },
  "Parameters": Object {
    "AMI": Object {
      "Default": "ami-6b50e218",
      "Description": "AMI ID",
      "Type": "String",
    },
    "AnghammaradSnsArn": Object {
      "Default": "/account/services/anghammarad.topic.arn",
      "Description": "SSM parameter containing the ARN of the Anghammarad SNS topic",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "InstanceType": Object {
      "Description": "EC2 instance type",
      "Type": "String",
    },
    "LoggingStreamName": Object {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "PackerInstanceProfile": Object {
      "Description": "Instance profile given to instances created by Packer",
      "Type": "String",
    },
    "PackerVersion": Object {
      "Description": "What version of Packer to install",
      "Type": "String",
    },
    "PrivateSubnets": Object {
      "Description": "Subnets to run the ASG and instances within",
      "Type": "List<AWS::EC2::Subnet::Id>",
    },
    "PublicSubnets": Object {
      "Description": "Subnets to run load balancer within",
      "Type": "List<AWS::EC2::Subnet::Id>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
    "TLSCert": Object {
      "Description": "ARN of a TLS certificate to install on the load balancer",
      "Type": "String",
    },
    "VPC": Object {
      "Description": "Virtual Private Cloud to run EC2 instances within",
      "Type": "AWS::EC2::VPC::Id",
    },
  },
  "Resources": Object {
    "AmigoAppPolicy": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sns:ListTopics",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "sns:*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Sub": Array [
                  "arn:aws:sns:*:*:amigo-\${Stage}-notify",
                  Object {
                    "Stage": Object {
                      "Ref": "Stage",
                    },
                  },
                ],
              },
            },
            Object {
              "Action": Array [
                "sns:*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Sub": Array [
                  "arn:aws:sns:*:*:amigo-\${Stage}-housekeeping-notify",
                  Object {
                    "Stage": Object {
                      "Ref": "Stage",
                    },
                  },
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetBucketPolicy",
                "s3:PutBucketPolicy",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:s3::*:",
                    Object {
                      "Ref": "DistributionBucketName",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "iam:GetInstanceProfile",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "PackerInstanceProfile",
              },
            },
          ],
        },
        "PolicyName": "amigo-app",
        "Roles": Array [
          Object {
            "Ref": "RootRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AmigoDataBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": Object {
          "Fn::FindInMap": Array [
            "stagemapping",
            Object {
              "Ref": "Stage",
            },
            "DataBucketName",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "AppPolicyF941AEC5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "AmigoDataBucket",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "dynamodb:ListTables",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "dynamodb:*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:*:*:table/amigo-",
                    Object {
                      "Ref": "Stage",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:DescribeTable",
                "dynamodb:GetItem",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:dynamodb:*:*:table/config-deploy",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "app-policy",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApplicationSecurityGroup": Object {
      "Properties": Object {
        "GroupDescription": "SSH and HTTP",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "10.249.0.0/16",
            "FromPort": 9000,
            "IpProtocol": "tcp",
            "ToPort": 9000,
          },
          Object {
            "FromPort": 9000,
            "IpProtocol": "tcp",
            "SourceSecurityGroupId": Object {
              "Ref": "LoadBalancerSecurityGroup",
            },
            "ToPort": 9000,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VPC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "AutoscalingGroup": Object {
      "Properties": Object {
        "AvailabilityZones": Object {
          "Fn::GetAZs": "",
        },
        "HealthCheckGracePeriod": 300,
        "HealthCheckType": "ELB",
        "LaunchConfigurationName": Object {
          "Ref": "LaunchConfig",
        },
        "LoadBalancerNames": Array [
          Object {
            "Ref": "LoadBalancer",
          },
        ],
        "MaxSize": "2",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VPCZoneIdentifier": Object {
          "Ref": "PrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "DescribeEC2PolicyFF5F9295": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablePolicyAmigoB25A5D2B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:s3:::",
                    Object {
                      "Ref": "DistributionBucketName",
                    },
                    "/deploy/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/amigo/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyAmigoB25A5D2B",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuAnghammaradSenderPolicy674A3874": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "AnghammaradSnsArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuAnghammaradSenderPolicy674A3874",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuLogShippingPolicy981BFE5A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:kinesis:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    Object {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceProfile": Object {
      "Properties": Object {
        "Path": "/",
        "Roles": Array [
          Object {
            "Ref": "RootRole",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "LaunchConfig": Object {
      "Properties": Object {
        "AssociatePublicIpAddress": false,
        "IamInstanceProfile": Object {
          "Ref": "InstanceProfile",
        },
        "ImageId": Object {
          "Ref": "AMI",
        },
        "InstanceType": Object {
          "Ref": "InstanceType",
        },
        "SecurityGroups": Array [
          Object {
            "Ref": "ApplicationSecurityGroup",
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Sub": Array [
              "#!/bin/bash -ev

wget -P /tmp https://releases.hashicorp.com/packer/\${PackerVersion}/packer_\${PackerVersion}_linux_arm64.zip
mkdir /opt/packer
unzip -d /opt/packer /tmp/packer_*_linux_arm64.zip
echo 'export PATH=\${!PATH}:/opt/packer' > /etc/profile.d/packer.sh

wget -P /tmp https://s3.amazonaws.com/session-manager-downloads/plugin/latest/ubuntu_arm64/session-manager-plugin.deb
dpkg -i /tmp/session-manager-plugin.deb

aws --region eu-west-1 s3 cp s3://\${DistributionBucketName}/deploy/\${Stage}/amigo/amigo_1.0-latest_all.deb /tmp/amigo.deb
dpkg -i /tmp/amigo.deb
",
              Object {
                "DistributionBucketName": Object {
                  "Ref": "DistributionBucketName",
                },
                "Stage": Object {
                  "Ref": "Stage",
                },
              },
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "LoadBalancer": Object {
      "Properties": Object {
        "CrossZone": true,
        "HealthCheck": Object {
          "HealthyThreshold": "2",
          "Interval": "10",
          "Target": "HTTP:9000/healthcheck",
          "Timeout": "5",
          "UnhealthyThreshold": "2",
        },
        "Listeners": Array [
          Object {
            "InstancePort": "9000",
            "LoadBalancerPort": "443",
            "Protocol": "HTTPS",
            "SSLCertificateId": Object {
              "Ref": "TLSCert",
            },
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Ref": "LoadBalancerSecurityGroup",
          },
        ],
        "Subnets": Object {
          "Ref": "PublicSubnets",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::ElasticLoadBalancing::LoadBalancer",
    },
    "LoadBalancerSecurityGroup": Object {
      "Properties": Object {
        "GroupDescription": "Guardian IP range has access to the load balancer on port 80",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "FromPort": 9000,
            "IpProtocol": "tcp",
            "ToPort": 9000,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "77.91.248.0/21",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VPC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "PackerPolicyE6040EA8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:AttachVolume",
                "ec2:AuthorizeSecurityGroupIngress",
                "ec2:CopyImage",
                "ec2:CreateImage",
                "ec2:CreateKeypair",
                "ec2:CreateSecurityGroup",
                "ec2:CreateSnapshot",
                "ec2:CreateTags",
                "ec2:CreateVolume",
                "ec2:DeleteKeypair",
                "ec2:DeleteSecurityGroup",
                "ec2:DeleteSnapshot",
                "ec2:DeleteVolume",
                "ec2:DeregisterImage",
                "ec2:DescribeImageAttribute",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "ec2:DescribeRegions",
                "ec2:DescribeSecurityGroups",
                "ec2:DescribeSnapshots",
                "ec2:DescribeSubnets",
                "ec2:DescribeTags",
                "ec2:DescribeVolumes",
                "ec2:DetachVolume",
                "ec2:GetPasswordData",
                "ec2:ModifyImageAttribute",
                "ec2:ModifyInstanceAttribute",
                "ec2:ModifySnapshotAttribute",
                "ec2:RegisterImage",
                "ec2:RunInstances",
                "ec2:StopInstances",
                "ec2:TerminateInstances",
                "iam:PassRole",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "packer-required-permissions",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PackerSecurityGroup": Object {
      "Properties": Object {
        "GroupDescription": "Security group for instances created by Packer",
        "GroupName": Object {
          "Fn::Sub": Array [
            "amigo-packer-\${Stage}",
            Object {
              "Stage": Object {
                "Ref": "Stage",
              },
            },
          ],
        },
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "FromPort": 0,
            "IpProtocol": "tcp",
            "ToPort": 65535,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VPC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "RootRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "ec2.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "amigo",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SSMRunCommandPolicy244E1613": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "ssm:StartSession",
                "ssm:TerminateSession",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-run-command-policy",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "RootRole",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
